### ISSUE TITLE
Multiple questions regarding the clocking signals

### ISSUE THREAD

**Issue by Algafix** [greatscottgadgets/hackrf#1091]
### Introduction

I was using the HackRF with a TCXO and the firmware 2018.01.1 to transmit a signal, but the RF signal produced was drifting. So I decided to output an RF sinusoidal signal at 40MHz and compare it with the one produced at the clock out port using an oscilloscope. Then I saw that the RF signal was not stable and was drifting all the time. In other words, when setting the trigger on the clock out signal, the RF signal was a blur: changing phase all the time.

### Exploring solutions

Therefore I decided to check the clocking signals produced by the Si5351C and realized that they are completely different that the ones indicated on the block diagram on this github:
* The CLK4 is transmitting to the RFFC5072 a 50MHz clocking signal
* The CLK7 is on.

So, assuming that this had to do with the firmware revision, I installed the 2021.03.1. Now, when I check the registers of the Si5351C, I see exactly the values indicated in the block diagram. But comparing the values with the 2018.01.1 I realize that:
* The CLK1 signal is Inverted
* The CLK4 signal is Inverted and at 40MHz.

### Questions

After exploring the documentation for the HackRF and the chips, I have some unresolved doubts:

* What is the reason behind inverting some clocking signals?
* Why was the RFFC5072 clock frequency changed from 50MHz to 40MHz?
  * Can I select an arbitrary frequency (assuming that the Si5351 can generate it)?
* Reading the lemondrop report on the signal phases I realized that maybe setting the MS4 as input for the CLK5 may help with the drifting problem (as both PLL would receive the same 40MHz signal), has someone tried that?

Thank you for the great work with HackRF.

**Comment by mossmann** [comment]
>     * What is the reason behind inverting some clocking signals?
>
>     * Why was the RFFC5072 clock frequency changed from 50MHz to 40MHz?

The CPLD clock (CLK1) was inverted to increase timing margin. The RFFC5072 reference clock (CLK4) was changed to 40 MHz in an effort to reduce spurious emissions (which can show up as spurs in RX mode). Reducing the number of different clock frequencies on the board also reduces the number of clock harmonics. It was inverted with the intent that it would have the opposite phase of the similar MAX2837 reference clock (CLK5) in hopes that this would reduce power consumption noise at the clock generator.

>       * Can I select an arbitrary frequency (assuming that the Si5351 can generate it)?

You could modify that clock frequency in firmware if the new frequency is supported by both the RFFC5072 and the Si5351C. You can see how such a modification was made in acaf0d1.

>     * Reading the lemondrop report on the signal phases I realized that maybe setting the MS4 as input for the CLK5 may help with the drifting problem (as both PLL would receive the same 40MHz signal), has someone tried that?

I actually thought that this was done in acaf0d1, but it looks like it was not. The intent was to have CLK4 and CLK5 always be perfectly out of phase, and I believe they must share a multisynth for that to be the case. If you try it, please let us know how it goes!

**Comment by Algafix** [comment]
Thank you for the answers and clarifications!

To discard other problems I did the same tests with an external much more precise clock generator instead of the TCXO. It took me a while since the clock signal was centered at 0V and the HackRF asks for 0V to 3.3V. But with a capacitor, 2 resistors and the Vcc pin I raised the signal from -1.6V/1.6V to 0V/3.3V.

My experiments with this external clock signal where precise enough, so I guess the problem was a faulty TCXO.

I did play with the RFFC5072 and its synths without getting any conclusive results. I had work to do and, once it worked out with the external clock, I did not progress further on that side. If in the future I see any improvement I'll let you know!

Thanks again!
